### PR TITLE
[QNEBE-682] CLI return code matches error output

### PR DESCRIPTION
**Pull request by cmvanb**
In order to return system status codes, this PR applies a refactor to the CLI error handling.

### Summary of changes:
`exceptions.py`:
- New exceptions added: `ApplicationFailedValidation`, `ExperimentFailedValidation`, `ExperimentExecutionError`.

`decorators.py`:
- Return status code 1 for predicted errors.
- Return status code 13 for unpredicted error.

`command_list.py`:
- Exceptions are raised for any type of explicit failure.
- `show_validation_messages` becomes `format_validation_messages`, the output of which is passed to the appropriate exception.

`test_command_list.py`:
- Test whether predictable status code 1 is returned.
- Update tests for changes in error message formatting.
- Fix shape of ErrorDict where it did not match it's type.

### CODE REVIEW

File: src/adk/command_list.py [hunk 1]
```diff
@@ -254,12 +258,12 @@ def applications_delete(
     deleted_completely = processor.applications_delete(application_name=application_name,
                                                        application_path=application_path)
     if deleted_completely:
-        typer.echo("Application deleted successfully")
+        typer.echo("Application deleted successfully.")
```
Comment (QFer) [inline src/adk/command_list.py:261]: It was a choice to not print the . (dot)
We also do not print them in the exceptions, now is it inconsistent
Some (not all) cli's we tried don't end their responses with end of line dots (dos, bash, kubectl), however others (git) do.

File: src/adk/command_list.py [hunk 2]
```diff
@@ -224,9 +228,9 @@ def applications_clone(
         validate_dict = processor.applications_validate(application_name=application_name,
                                                         application_path=application_path)
         if validate_dict["error"] or validate_dict["warning"]:
-            show_validation_messages(validate_dict)
-            typer.echo(f"Local application '{application_name}' is invalid. Application not cloned")
-            return
+            typer.echo("Local application cannot be cloned.")
```
Comment (QFer) [inline src/adk/command_list.py:231]: This changes the message a bit. The app can be cloned (more or less a copy) when it is validated incorrectly, but we didn't clone it on purpose. We chose only to clone correctly validated apps (after the clone the app is then also valid).

File: src/tests/test_command_list.py [hunk 3]
```diff
@@ -244,7 +240,7 @@ def test_applications_local_clone_success(self):
             application_exists_mock.return_value = False, ""
             retrieve_appname_and_path_mock.return_value = self.path, self.application
             # When application is valid (no items in error, warning and info)
-            applications_validate_mock.return_value = {"error": {}, "warning": {}, "info": {}}
+            applications_validate_mock.return_value = {"error": [], "warning": [], "info": []}
```
Comment (QFer) [inline src/tests/test_command_list.py:243]: Nice catch!

File: src/tests/test_command_list.py [hunk 4]
```diff
@@ -312,10 +308,12 @@ def test_applications_clone_exceptions(self):
             application_exists_mock.return_value = False, ""
             retrieve_appname_and_path_mock.return_value = self.path, self.application
             # When application is valid (no items in error, warning and info)
-            applications_validate_mock.return_value = {"error": {"an_error"}, "warning": {}, "info": {}}
+            applications_validate_mock.return_value = {"error": [ "an_error" ], "warning": [], "info": []}
```
Comment (QFer) [inline src/tests/test_command_list.py:311]: this is not PEP8.  ["an_error"] is PEP8. Also counts for line 424 and the other instances where it is changed

File: src/adk/command_list.py [hunk 5]
```diff
@@ -278,16 +282,18 @@ def applications_upload(
     application_path, application_name = retrieve_application_name_and_path(application_name=application_name)
     validate_dict = processor.applications_validate(application_name=application_name,
                                                     application_path=application_path)
+    validation_messages = format_validation_messages(validate_dict)
+
     if validate_dict["error"] or validate_dict["warning"]:
-        show_validation_messages(validate_dict)
-        typer.echo(f"Application '{application_name}' is invalid. Application not uploaded")
+        typer.echo("Application could not be uploaded.")
```
Comment (QFer) [inline src/adk/command_list.py:288]: Also here the message is changed a bit. We can but we do not upload invalid apps

File: src/tests/test_command_list.py [hunk 6]
```diff
@@ -968,12 +970,12 @@ def test_experiment_run_update_fails(self):
...
             exp_run_output = self.runner.invoke(experiments_app, ['run', '--timeout=30', '--update'])
             exp_run_mock.assert_not_called()
-            show_validation_messages_mock.assert_called_once()
+            format_validation_messages_mock.assert_called_once()
             retrieve_expname_and_path_mock.assert_called_once()
             retrieve_appname_and_path_mock.assert_called_once()
-            self.assertEqual(exp_run_output.exit_code, 0)
-            self.assertIn(f"Application '{self.application}' is invalid. Experiment cannot be updated.",
-                          exp_run_output.stdout)
+            self.assertEqual(exp_run_output.exit_code, 1)
```
Comment (QFer) [inline src/tests/test_command_list.py:976]: Maybe add a test where we get the 13 as error code.

**Review by QFer — CHANGES_REQUESTED** [review]

File: src/adk/command_list.py (hunk 1 above)
Comment (cmvanb) [reply src/adk/command_list.py:261]: Good grammar is good grammar regardless of who else does it... but I take your point, it should be consistent.

File: src/adk/command_list.py (hunk 2 above)
Comment (cmvanb) [reply src/adk/command_list.py:231]: Ok, I will fix these messages.

File: src/tests/test_command_list.py (hunk 6 above)
Comment (QFer) [reply src/tests/test_command_list.py:976]: Maybe not.

**Review by QFer — APPROVED** [review]

**Pull request merged**